File: guests/swift/menus.swift
// The menus scene, Swift port — guests/rust/menus.rs,
// tools/scenes/menus.steps.

import Foundation

let app = KayaApp()

let (groups, items) = app.build { tx -> (KayaCollection, KayaCollection) in
    let status = tx.signal(.str("ready"))
    let canExport = tx.signal(.bool(false))
    let details = tx.signal(.bool(false))
    let sort = tx.signal(.f64(0.0))

    let onShare: (KayaAppTx) throws -> Void = { t in t.write(status, .str("shared")) }

    let share = tx.item("Share", primary: true, onActivate: onShare)
    let file = tx.menu(
        "File", enabled: canExport,
        items: [
            // No `save` in the symbol vocabulary; `done` is the checkmark
            // idiom (docs/styling-plan.md D6).
            tx.item("Save", shortcut: "primary+s", symbol: .done) { t in
                t.write(status, .str("saved"))
            },
            tx.item("Export", enabled: canExport, symbol: .forward),
            share,
        ])
    let view = tx.menu(
        "View",
        items: [
            tx.toggle("Details", checked: details, symbol: .info) { t, on in
                t.write(status, .str(on ? "details on" : "details off"))
            }
        ])
    // Option order IS the index vocabulary: Name = 0, Date = 1.
    let sortGroup = tx.radioGroup(
        "Sort", options: [tx.option("Name"), tx.option("Date")], value: sort
    ) { t, index in
        t.write(status, .str(index == 1 ? "sorted date" : "sorted name"))
    }
    tx.window(title: "menus", menus: [file, view, sortGroup])

    let groups = tx.collection()
    var itemsOut: KayaCollection!
    let catalog = tx.contextCatalog(items: [
        tx.item("Remove", symbol: .delete) { t, keys in
            guard case .str(let group) = keys[0], case .str(let item) = keys[1] else {
                return
            }
            t.remove(itemsOut.at(.str(group)), .str(item))
            t.write(status, .str("removed \(group)/\(item)"))
        }
    ])

    let root = tx.column {
        tx.label(bind: status)  // label#0
        tx.button("enable export") { t in  // button#0
            t.write(canExport, .bool(true))
        }
        tx.button("reset menu state") { t in  // button#1
            t.write(details, .bool(false))
            t.write(sort, .f64(0.0))
            t.write(status, .str("ready"))
        }
        tx.button("extend menus") { t in  // button#2
            t.menu(share, primary: false)
            t.menu(
                file, label: "Document",
                items: [
                    t.item("Publish", symbol: .copy, primary: true, onActivate: onShare)
                ])
            t.window(menus: [t.menu("Tools", items: [t.item("Inspect", symbol: .search)])])
        }

        let targetText = tx.signal(.str("rename target"))
        let target = tx.label(bind: targetText)  // label#1
        tx.contextMenu(
            target,
            items: [
                tx.item("Rename", symbol: .edit) { t in
                    t.write(status, .str("renamed"))
                }
            ])

        tx.each(groups) { g in
            let items = g.collection()
            itemsOut = items
            // A bare mention in the builder is DISCARDED, so the For is
            // declared INSIDE the column (docs/traps.md, result builders).
            _ = g.column {
                g.each(items) { r in
                    // label#2 once g2/a stamps.
                    let row = r.label(KayaField<String>.element)
                    r.contextMenu(row, catalog)
                }
            }
        }
    }
    tx.mount(root)
    return (groups, itemsOut)
}

// Seeded after the mount, so the copy stamps from a closed template.
app.build { tx in
    tx.insert(groups, .str("g2"), .str("Home"))
    tx.insert(items.at(.str("g2")), .str("a"), .str("water plants"))
}

app.run()

File: guests/ocaml/menus.ml
(* The menus scene, OCaml port — guests/rust/menus.rs,
   tools/scenes/menus.steps. *)

open Kaya_wire
open Kaya_app

let () =
  let app = Kaya_app.create () in

  let groups, items =
    build app (fun () ->
       let status = signal (Str "ready") in
       let can_export = signal (Bool false) in
       let details = signal (Bool false) in
       let sort = signal (F64 0.0) in

       let on_share () = write status (Str "shared") in

       let share = item ~label:"Share" ~primary:true ~on_activate:on_share () in
       let file =
         menu ~label:"File" ~bind_enabled:can_export
           [
             (* No `save` in the symbol vocabulary; [Done] is the checkmark
                idiom (docs/styling-plan.md D6). *)
             item ~label:"Save" ~symbol:Done ~shortcut:"primary+s"
               ~on_activate:(fun () -> write status (Str "saved"));
             item ~label:"Export" ~bind_enabled:can_export ~symbol:Forward;
             w share;
           ]
           ()
       in
       window ~title:"menus"
         ~menus:
           [
             w file;
             menu ~label:"View"
               [
                 toggle ~label:"Details" ~bind_checked:details ~symbol:Info
                   ~on_toggle:(fun on ->
                     write status
                       (Str (if on then "details on" else "details off")));
               ];
             (* Option order IS the index vocabulary: Name = 0, Date = 1. *)
             radio_group ~label:"Sort" ~bind_value:sort
               ~on_select:(fun index ->
                 write status
                   (Str (if index = 1 then "sorted date" else "sorted name")))
               [ option ~label:"Name"; option ~label:"Date" ];
           ]
         ();

       let groups = collection () in
       let items_ref = ref None in
       (* One catalog shared across every stamped copy. [items_ref] is how
          the handler reaches a collection [for_each] has not returned yet. *)
       let catalog =
         context_catalog
           [
             item ~label:"Remove" ~symbol:Delete
               ~on_activate_node:(fun keys ->
                 match keys with
                 | [ Str group; Str item ] ->
                     remove (at (Option.get !items_ref) (Str group)) (Str item);
                     write status
                       (Str (Printf.sprintf "removed %s/%s" group item))
                 | _ -> ());
           ]
       in

       let group_list, items =
         for_each groups
           (fun () ->
             Tpl.(
               let items = collection () in
               let _ =
                 column
                   [
                     each items (fun () ->
                         (* label#2 once g2/a stamps. *)
                         let row = label ~bind_field:element () in
                         context_menu row catalog);
                   ]
                   ()
               in
               items))
           ()
       in
       items_ref := Some items;

       let target_text = signal (Str "rename target") in
       let root =
         column
           [
             label ~bind:status (* label#0 *);
             button ~text:"enable export"
               ~on_click:(fun () -> write can_export (Bool true)) (* button#0 *);
             button ~text:"reset menu state"
               ~on_click:(fun () ->
                 write details (Bool false);
                 write sort (F64 0.0);
                 write status (Str "ready")) (* button#1 *);
             button ~text:"extend menus"
               ~on_click:(fun () ->
                 set_menu_primary share false;
                 set_menu_label file "Document";
                 menu_append file
                   [
                     item ~label:"Publish" ~primary:true ~symbol:Copy
                       ~on_activate:on_share;
                   ];
                 window
                   ~menus:
                     [
                       menu ~label:"Tools"
                         [ item ~label:"Inspect" ~symbol:Search ];
                     ]
                   ()) (* button#2 *);
             (fun () ->
               let target = label ~bind:target_text () (* label#1 *) in
               context_menu target
                 [
                   item ~label:"Rename" ~symbol:Edit
                     ~on_activate:(fun () -> write status (Str "renamed"));
                 ];
               target);
             w group_list;
           ]
           ()
       in
       mount root;
       (groups, items))
  in

  (* Seeded after the mount, so the copy stamps from a closed template. *)
  build app (fun () ->
     insert groups (Str "g2") (Str "Home");
     insert (at items (Str "g2")) (Str "a") (Str "water plants"));

  exit (run app)
